kms/main.tf

```diff
@@ -0,0 +1,168 @@
+    }
+    actions   = ["kms:DescribeKey", "kms:GetKeyPolicy"]
+    resources = ["*"]
+  }
+  statement {
+    sid = "AdminRoleAdministerKey"
+    principals {
+      type        = "AWS"
+      identifiers = [module.kms_admin_role.role_arn]
+    }
+    actions = [
+      "kms:Create*",
+      "kms:Describe*",
+      "kms:Enable*",
+      "kms:List*",
+      "kms:Put*",
+      "kms:Update*",
+      "kms:Revoke*",
+      "kms:Disable*",
+      "kms:Get*",
+      "kms:Delete*",
+      "kms:TagResource",
+      "kms:UntagResource",
+      "kms:ScheduleKeyDeletion",
+      "kms:CancelKeyDeletion"
+    ]
+    resources = ["*"]
+  }
+  dynamic "statement" {
+    for_each = length(var.default_policy_variables.ci_roles) == 0 ? [] : ["ci_roles"]
+    content {
+      sid = "CIRolesAdministerWithoutDelete"
+      principals {
+        type        = "AWS"
+        identifiers = var.default_policy_variables.ci_roles
+      }
+      actions = [
+        "kms:Create*",
+        "kms:Describe*",
+        "kms:Enable*",
+        "kms:List*",
+        "kms:Put*",
+        "kms:Update*",
+        "kms:Revoke*",
+        "kms:Disable*",
+        "kms:Get*",
+        "kms:TagResource",
+        "kms:UntagResource"
+      ]
+      resources = ["*"]
+    }
+  }
+  dynamic "statement" {
+    for_each = length(var.default_policy_variables.user_roles) == 0 ? [] : ["user_roles"]
+    content {
+      sid = "UserRolesEncryptAndDecrypt"
+      principals {
+        type        = "AWS"
+        identifiers = var.default_policy_variables.user_roles
+      }
+      actions = [
+        "kms:Encrypt",
+        "kms:Decrypt",
+        "kms:ReEncrypt*",
+        "kms:GenerateDataKey*",
+        "kms:DescribeKey"
+      ]
+      resources = ["*"]
+    }
+  }
+  dynamic "statement" {
+    for_each = length(var.default_policy_variables.persistent_resource_roles) == 0 ? [] : ["persistent_resource_roles"]
+    content {
+      sid = "ResourceRolesGrants"
+      principals {
+        type        = "AWS"
+        identifiers = var.default_policy_variables.persistent_resource_roles
+      }
+      actions = [
+        "kms:CreateGrant",
+        "kms:ListGrants",
+        "kms:RevokeGrant"
+      ]
+      resources = ["*"]
+      condition {
+        test     = "Bool"
+        values   = ["kms:GrantIsForAWSResource"]
+        variable = "true"
+      }
+    }
+  }
+  dynamic "statement" {
+    for_each = var.default_policy_variables.service_names
+    content {
+      sid = "AllowSameAccountServiceAccess${title(statement.value)}"
+      principals {
+        type        = "Service"
+        identifiers = ["${statement.value}.amazonaws.com"]
+      }
+      actions = [
+        "kms:Decrypt",
+        "kms:GenerateDataKey*",
+        "kms:DescribeKey"
+      ]
+      effect    = "Allow"
+      resources = ["*"]
+      condition {
+        test     = "StringEquals"
+        values   = [data.aws_caller_identity.current.account_id]
+        variable = "aws:SourceAccount"
+      }
+    }
```

Comment:
Will we need to do something special here for cross account access? E.g. subscribing an SQS to an encrypted SNS in another account?

Reply:
I believe you can just change the source account which should allow services from other accounts.I've added a variable for source account which defaults to the calling account if it's not provided.